### DISCUSSION TITLE
중복된 이메일로 회원 가입을 할 경우, 409와 422 상태코드 중 어떤 것이 더 적절할까?

### DISCUSSION THREAD

**Discussion by devjun10** [dailyge/dailyge-server#29] [Decision]
## 📝 연관 PR
- [[DLG-07] User 등록 API를 개발한다.](https://github.com/dailyge/dailyge-server/pull/3#top)

<br/><br/>

## ❓ 작성 이유

중복된 이메일로 회원 가입을 할 경우 어떠한 상태코드를 내려주는 것이 좋을 지 협의가 필요하였습니다.

<br/><br/>

## 409를 사용하자는 의견
 - 기존에 존재하는 리소스와 충돌이 발생했을 경우 사용하는 상태코드이다.
 - 이 리소스를 이메일로 본다면 리소스가 충돌한 것으로 볼 수 있다.
 - [ID 중복체크](https://www.webfx.com/web-development/glossary/http-status-codes/what-is-a-409-status-code/) 예시로 409를 사용하는 글도 존재한다.

<br/><br/>

## 422를 사용하자는 의견
 - 요청은 올바르나 이를 처리할 수 없는 422 응답을 내리는게 조금 더 적절할 것 같다.
 - [422 응답코드 설명](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422)
 - [스택오버플로우](https://stackoverflow.com/questions/3825990/http-response-code-for-post-when-resource-already-exists)에서의 글도 보면 조금 다른 느낌이다.

<br/><br/>

## 최종 결정은 422.
> 요청은 올바르나 이를 처리할 수 없는 422 응답을 내리는게 조금 더 적절할 것 같다.

<br/>

**Comment by kmularise** [comment]
글 잘봤습니다. 409는 클라이언트와 요청한 상태와 서버의 현재 상태가 충돌할 때, 즉 리소스 상태가 충돌할 때 사용하고 422는 요청은 올바르나 이를 처리할 수 없을 때 사용하는 군요! 쿠폰 당첨자 선정 API를 개발할 때, 이미 실행된 작업임을 나타내는 에러 상태코드를 결정할 때 도움이 될 것 같아요!
